ibmcloud/image/build.sh

```diff
@@ -206,69 +254,165 @@ umount "$src_mnt/dev"
+    chroot "$dst_mnt" update-initramfs -u || true
+    echo "!!! Bootloader install errors prior to this line are intentional !!!!!" 1>&2
+    printf "\nGenerating an IBM Secure Execution image\n"
+    # Clean up kernel names and make sure they are where we expect them
+    KERNEL_FILE=$(readlink ${dst_mnt}/boot/vmlinuz)
+    echo "using kernel: ${KERNEL_FILE}"
+    INITRD_FILE=$(readlink ${dst_mnt}/boot/initrd.img)
+    echo "using initrd: ${INITRD_FILE}"
+    SE_PARMLINE="root=/dev/mapper/$LUKS_NAME console=ttysclp0 quiet panic=0 rd.shell=0 blacklist=virtio_rng swiotlb=262144"
+    echo "${SE_PARMLINE}" > ${dst_mnt}/boot/parmfile
+    host_keys=""
+    for i in $(ls ${HOST_KEYS_DIR}/*.crt); do
+        echo "found host key file: \"${i}\""
+        host_keys+="-k ${i} "
+    done
+    /usr/bin/genprotimg \
+        -i ${dst_mnt}/boot/${KERNEL_FILE} \
+        -r ${dst_mnt}/boot/${INITRD_FILE} \
+        -p ${dst_mnt}/boot/parmfile \
+        --no-verify \
+        ${host_keys} \
+        -o "$dst_mnt"/boot-se/se.img
+    # exit and throw an error if no se image was created
+    [ ! -e $dst_mnt/boot-se/se.img ] && exit 1
+    # if building the image succeeded wipe /boot
+    rm -rf ${dst_mnt}/boot/*
+    printf "\nRunning zipl to prepare boot partition\n"
+    chroot $dst_mnt zipl --targetbase $tmp_nbd \
+    --targettype scsi \
+    --targetblocksize 512 \
+    --targetoffset 2048 \
+    --target /boot-se \
+    --image /boot-se/se.img
+    printf "\nClean luks keyfile\n"
+    shred -u ./rootkeys/rootkey.bin
+    umount ./rootkeys/
+    rm -rf ./rootkeys/
+    umount $dst_mnt/etc/keys
+    umount $dst_mnt/boot-se
+else
+    case "$(uname -m)" in
+        s390x)
+            printf "\nExecuting zipl"
+            helper="$dst_mnt/lib/s390-tools/zipl_helper.nbd"
+            cat <<END > "$helper"
 echo "targetbase=$tmp_nbd"
 echo "targettype=scsi"
 echo "targetblocksize=512"
 echo "targetoffset=$target_offset"
 END
-        chmod 755 "$helper"
-        chroot "$dst_mnt" zipl -V
-        rm "$helper"
-        ;;
-    *)
-        echo -e "\nUpdating GRUB settings\n"
-        sed -i -r -e 's|^GRUB_CMDLINE_LINUX=|#\0"|' "$dst_mnt/etc/default/grub"
-        cat <<END >> "$dst_mnt/etc/default/grub"
+            chmod 755 "$helper"
+            chroot "$dst_mnt" zipl -V
+            rm "$helper"
+            ;;
+        *)
+            printf "\nUpdating GRUB settings"
+            sed -i -r -e 's|^GRUB_CMDLINE_LINUX=|#\0"|' "$dst_mnt/etc/default/grub"
+            cat <<END >> "$dst_mnt/etc/default/grub"
 
 GRUB_CMDLINE_LINUX="nomodeset nofb vga=normal console=ttyS0"
 GRUB_DISABLE_LINUX_UUID=true
 GRUB_DISABLE_OS_PROBER=true
 GRUB_DEVICE="LABEL=cloudimg-rootfs"
 END
-        chroot "$dst_mnt" update-grub
-        ;;
-esac
+            chroot "$dst_mnt" update-grub
+            ;;
+    esac
+fi
 
 umount "$dst_mnt/dev"
 umount "$dst_mnt/proc"
 umount "$dst_mnt/sys"
-
 umount "$dst_mnt"
 
+if [ "${SE_BOOT}" = "1" ]; then
```

Comment:
Any reason to not leave this to the `cleanup` function called on L413?

Reply:
need close encrypted root partition before `qemu-nbd --disconnect`
the cleanup function is the final step.